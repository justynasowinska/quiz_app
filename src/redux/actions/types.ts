// questions
export const UPDATE_QUESTIONS = 'update_questions';

// progress
export const SET_NEXT_QUESTION = 'set_next_question';
export const RESET_PROGRESS = 'reset_progress';

// answers
export const ADD_USER_ANSWER = 'add_user_answer';
export const RESET_ANSWERS = 'reset_answers';