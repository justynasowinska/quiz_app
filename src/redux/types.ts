
export type DifficultlyType = 'hard' | 'medium' | 'easy';
export type BooleanCorrectAnswerType = 'True' | 'False';

export interface QuizQuestionType {
    category: string;
    type: string;
    difficulty: DifficultlyType;
    question: string;
    correct_answer: BooleanCorrectAnswerType; // maybe just string?
    incorrect_answers: Array<BooleanCorrectAnswerType>;
}

export type QuizQuestionsStateType = Array<QuizQuestionType>;

export interface QuizProgressType {
    currentQuestion: number;
}

export interface QuizAnswerType {
    questionId: number;
    answer: BooleanCorrectAnswerType;
}

export type QuizAnswersType = Array<QuizAnswerType>;