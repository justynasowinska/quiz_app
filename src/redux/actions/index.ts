export * from './answers';
export * from './progress';
export * from './questions';