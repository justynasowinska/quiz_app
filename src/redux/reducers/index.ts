import { combineReducers } from 'redux';

import quizDataReducer from './quizQuestionsReducer';
import quizProgressReducer from './quizProgressReducer';
import quizAnswersReducer from './quizAnswersReducer';

export const rootReducer = combineReducers({
    questions: quizDataReducer,
    progress: quizProgressReducer,
    answers: quizAnswersReducer
});

export default rootReducer;