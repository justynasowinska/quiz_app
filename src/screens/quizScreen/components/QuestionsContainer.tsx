import * as React from 'react';
import { Text, StyleSheet, View } from 'react-native';

import { BooleanQuestionsDeck } from './BooleanQuestionsDeck';
import { QuizQuestionType, BooleanCorrectAnswerType } from '../../../redux/types';
import { getDifficultyLevelColor } from '../../../utils/helpers';

interface PropsType {
    questions: ReadonlyArray<QuizQuestionType>;
    currentQuestionIndex: number;
    onUserAnswer: (questionIndex: number, answer: BooleanCorrectAnswerType) => void;
}

export const QuestionsContainer = (props: PropsType) => {
    const { questions, currentQuestionIndex, onUserAnswer } = props;
    const difficulty = questions[currentQuestionIndex].difficulty;
    const difficultyColor = getDifficultyLevelColor(difficulty);

    const getCurrentProgress = () => {
        const { questions, currentQuestionIndex } = props;

        return `${currentQuestionIndex} / ${questions.length}`;
    };

    // TODO: Render other questions deck if not boolean questions type
    return (
        <View style={styles.container}>
            <Text style={[styles.levelText, { color: difficultyColor }]}>
                Level: {difficulty}
            </Text>
            <BooleanQuestionsDeck
                questions={questions}
                currentQuestionIndex={currentQuestionIndex}
                onUserAnswer={onUserAnswer}
            />
            <Text style={styles.progressText}>
                {getCurrentProgress()}
            </Text>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
    },
    levelText: {
        fontSize: 15,
        marginBottom: 2,
        alignSelf: 'flex-start'
    },
    progressText: {
        fontSize: 15,
        marginTop: 2
    }
});