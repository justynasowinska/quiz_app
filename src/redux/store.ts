import { createStore, applyMiddleware } from 'redux';
import { persistStore, persistReducer } from 'redux-persist';
import autoMergeLevel2 from 'redux-persist/es/stateReconciler/autoMergeLevel2';
import storage from 'redux-persist/lib/storage';

import reducers from './reducers';

const persistConfig = {
    key: 'root',
    storage,
    stateReconciler: autoMergeLevel2
};

const persistedReducer = persistReducer(persistConfig, reducers);

const middlewares = [];

if (__DEV__) {
    const { logger } = require('redux-logger');

    middlewares.push(logger);
}

export const store = createStore(persistedReducer, applyMiddleware(...middlewares));
export const persistor = persistStore(store);