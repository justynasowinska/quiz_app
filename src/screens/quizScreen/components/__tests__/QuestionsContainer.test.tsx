import * as React from 'react';
import { shallow } from 'enzyme';

import { QuestionsContainer } from '../QuestionsContainer';
import { QuizQuestionType } from '../../../../redux/types';

describe('Quiz Screen QuestionsContainer Component', () => {
    it('matches to the snapshot', () => {
        const questions: ReadonlyArray<QuizQuestionType> = [{
            category: 'Entertainment: Video Games',
            type: 'boolean',
            difficulty: 'hard',
            question: 'Unturned originally started as a Roblox game.',
            correct_answer: 'True',
            incorrect_answers: ['False']
        }];

        const component = shallow(
            <QuestionsContainer
                questions={questions}
                currentQuestionIndex={0}
                onUserAnswer={jest.fn()}
            />
        );

        expect(component).toMatchSnapshot();
    });
});
