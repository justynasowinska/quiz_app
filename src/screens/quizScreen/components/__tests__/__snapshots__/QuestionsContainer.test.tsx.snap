// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz Screen QuestionsContainer Component matches to the snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "alignSelf": "flex-start",
          "fontSize": 15,
          "marginBottom": 2,
        },
        Object {
          "color": "#CD0720",
        },
      ]
    }
  >
    Level: 
    hard
  </Text>
  <BooleanQuestionsDeck
    currentQuestionIndex={0}
    onUserAnswer={[MockFunction]}
    questions={
      Array [
        Object {
          "category": "Entertainment: Video Games",
          "correct_answer": "True",
          "difficulty": "hard",
          "incorrect_answers": Array [
            "False",
          ],
          "question": "Unturned originally started as a Roblox game.",
          "type": "boolean",
        },
      ]
    }
  />
  <Text
    style={
      Object {
        "fontSize": 15,
        "marginTop": 2,
      }
    }
  >
    0 / 1
  </Text>
</View>
`;
